File: dart/test/semantic_index_runtime_projection_test.dart
// FUTURE-PARITY-BACKLOG.10.5.5.2 — immutable observed-index derivation.

import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:linkedspec_dart/src/semantic/sha256.dart';
import 'package:test/test.dart';

const _input = 'ab\n';
const _inputIdentity =
    'input:sha256:a63d8014dba891345b30174df2b2a57efbb65b4f9f09b98f245d1b3192277ece';
const _responseDigest =
    '36897041c6f71b95b577ce7b38f42d3649c6adffc6c37c069944a90f6eb65887';

void main() {
  late String source;
  late CompiledSpec compiled;
  late SemanticIndex base;

  setUpAll(() {
    source = File(
      '../capability_conformance/semantic_introspection/runtime.spec',
    ).readAsStringSync();
    final parsed = parseSpec(source);
    validateSpec(parsed);
    compiled = compileSpec(parsed);
    base = SemanticIndex.fromSource(
      source,
      options: const SemanticIndexOptions(
        logicalName: 'runtime.spec',
        sourceDetailCeiling: SemanticSourceDetail.text,
      ),
    );
  });

  test('derives an immutable snapshot matching the twentieth digest', () {
    final events = _capture(compiled);
    final request = SemanticQuery(
      operation: SemanticQueryOperation.list,
      recordKinds: const ['execution', 'event'],
      source: const SemanticQuerySource(detail: SemanticSourceDetail.identity),
    );
    final neutralRequest = request.toJson();

    expect(base.snapshot.hasExecution, isFalse);
    expect(base.query(request).records, isEmpty);
    final derived = base.withExecutionObservation(events);
    expect(derived.snapshot.hasExecution, isTrue);
    final typed = derived.query(request);
    final neutral = derived.queryNeutral(neutralRequest);

    expect(typed, neutral);
    expect(typed.records.map((record) => record.id), [
      'execution:0',
      'event:execution:0:0',
      'event:execution:0:1',
      'event:execution:0:2',
    ]);
    expect(_canonicalDigest(typed.toJson()), _responseDigest);

    final relations = derived.query(
      SemanticQuery(
        operation: SemanticQueryOperation.relations,
        subjects: const ['execution:0'],
        relationKinds: const ['observed_as'],
        source: const SemanticQuerySource(
          detail: SemanticSourceDetail.identity,
        ),
      ),
    );
    expect(relations.relations.map((relation) => relation.id), [
      'relation:observed_as:execution:0:event:execution:0:0:0',
      'relation:observed_as:execution:0:event:execution:0:1:1',
      'relation:observed_as:execution:0:event:execution:0:2:2',
    ]);
    expect(relations.relations.map((relation) => relation.evidenceIds), [
      ['regex:rule:Top:0'],
      ['regex:rule:Top:1'],
      ['rule:Top'],
    ]);

    events[0] = _slot(position: 999);
    final detached = typed.toJson();
    final records = detached['records']! as List<Object?>;
    final execution = records.first as Map<String, Object?>;
    (execution['facts']! as Map<String, Object?>)['status'] = 'mutated';
    expect(_canonicalDigest(derived.query(request).toJson()), _responseDigest);
    expect(base.snapshot.hasExecution, isFalse);
    expect(base.query(request).records, isEmpty);
  });

  test('rejects malformed foreign reordered and duplicate-final events', () {
    final valid = _expectedEvents();
    final cases = <(String, List<RuntimeSemanticObservationEvent>)>[
      ('empty', []),
      ('missing result', valid.take(2).toList()),
      ('duplicate final', [...valid, valid.last]),
      ('reordered final', [valid[0], valid[2], valid[1]]),
      (
        'foreign contract',
        [_slot(contractId: 'future-contract'), valid[1], valid[2]],
      ),
      ('missing rule label', [_slot(ruleLabel: ''), valid[1], valid[2]]),
      ('missing slot target', [_slot(targetRule: null), valid[1], valid[2]]),
      ('missing slot index', [_slot(regexIndex: null), valid[1], valid[2]]),
      ('negative slot index', [_slot(regexIndex: -1), valid[1], valid[2]]),
      ('negative slot position', [_slot(position: -1), valid[1], valid[2]]),
      (
        'slot carries result state',
        [_slot(status: 'succeeded'), valid[1], valid[2]],
      ),
      (
        'result carries slot state',
        [valid[0], valid[1], _result(targetRule: 'Top', regexIndex: 0)],
      ),
      ('foreign slot', [_slot(targetRule: 'Missing'), valid[1], valid[2]]),
      (
        'foreign result rule',
        [valid[0], valid[1], _result(ruleLabel: 'Missing')],
      ),
      ('failed result', [valid[0], valid[1], _result(status: 'failed')]),
      ('negative result position', [valid[0], valid[1], _result(position: -1)]),
      (
        'missing input identity',
        [valid[0], valid[1], _result(inputIdentity: null)],
      ),
      (
        'malformed input identity',
        [valid[0], valid[1], _result(inputIdentity: 'input:sha256:xyz')],
      ),
    ];

    for (final (label, observation) in cases) {
      expect(
        () => base.withExecutionObservation(observation),
        throwsA(
          isA<SemanticIndexError>()
              .having((error) => error.stage, 'stage', 'execution_observation')
              .having(
                (error) => error.code,
                'code',
                'semantic_index_invalid_observation',
              ),
        ),
        reason: label,
      );
    }
  });

  test('requires the selecting rule to own the selected slot relation', () {
    const unrelatedSource = '''
Top::
 /a/ -> Top[0] { return("top") }

Other::
 /b/ -> Other[0] { return("other") }
''';
    final unrelated = SemanticIndex.fromSource(
      unrelatedSource,
      options: const SemanticIndexOptions(
        logicalName: 'unrelated.spec',
        sourceDetailCeiling: SemanticSourceDetail.text,
      ),
    );
    final events = [
      _slot(ruleLabel: 'Top', targetRule: 'Other'),
      _result(ruleLabel: 'Top'),
    ];

    expect(
      () => unrelated.withExecutionObservation(events),
      throwsA(
        isA<SemanticIndexError>().having(
          (error) => error.message,
          'message',
          contains("does not select regex slot 'Other[0]'"),
        ),
      ),
    );
  });

  test('an already observed or failed index cannot derive again', () {
    final valid = _expectedEvents();
    final derived = base.withExecutionObservation(valid);
    expect(
      () => derived.withExecutionObservation(valid),
      throwsA(
        isA<SemanticIndexError>()
            .having((error) => error.stage, 'stage', 'execution_observation')
            .having(
              (error) => error.code,
              'code',
              'semantic_index_invalid_observation',
            ),
      ),
    );

    final failed = SemanticIndex.fromSource(
      'Top:: Missing\n',
      options: const SemanticIndexOptions(
        logicalName: 'failed.spec',
        sourceDetailCeiling: SemanticSourceDetail.span,
      ),
    );
    expect(failed.snapshot.state, SemanticSnapshotState.failedCompilation);
    expect(
      () => failed.withExecutionObservation(valid),
      throwsA(isA<SemanticIndexError>()),
    );
  });
}

List<RuntimeSemanticObservationEvent> _capture(CompiledSpec compiled) {
  final events = <RuntimeSemanticObservationEvent>[];
  final result = LinkedSpecRuntimeEngine(
    compiled,
  ).parse(_input, semanticObservationSink: events.add);
  expect(result.value, ['A', 'B']);
  return events;
}

List<RuntimeSemanticObservationEvent> _expectedEvents() => [
  _slot(),
  _slot(regexIndex: 1, position: 2),
  _result(),
];

RuntimeSemanticObservationEvent _slot({
  String contractId = linkedSpecSemanticExecutionObservationContract,
  String ruleLabel = 'Top',
  String? targetRule = 'Top',
  int? regexIndex = 0,
  int position = 1,
  String? inputIdentity,
  String? status,
}) => RuntimeSemanticObservationEvent(
  contractId: contractId,
  eventKind: RuntimeSemanticObservationEventKind.regexSlotSelected,
  ruleLabel: ruleLabel,
  targetRule: targetRule,
  regexIndex: regexIndex,
  position: position,
  inputIdentity: inputIdentity,
  status: status,
);

RuntimeSemanticObservationEvent _result({
  String contractId = linkedSpecSemanticExecutionObservationContract,
  String ruleLabel = 'Top',
  String? targetRule,
  int? regexIndex,
  int position = 2,
  String? inputIdentity = _inputIdentity,
  String? status = 'succeeded',
}) => RuntimeSemanticObservationEvent(
  contractId: contractId,
  eventKind: RuntimeSemanticObservationEventKind.ruleResult,
  ruleLabel: ruleLabel,
  targetRule: targetRule,
  regexIndex: regexIndex,
  position: position,
  inputIdentity: inputIdentity,
  status: status,
);

String _canonicalDigest(Map<String, Object?> value) =>
    sha256Hex(utf8.encode(jsonEncode(_canonicalValue(value))));

Object? _canonicalValue(Object? value) {
  if (value case Map<Object?, Object?>()) {
    final keys = value.keys.map((key) => key.toString()).toList()..sort();
    return <String, Object?>{
      for (final key in keys) key: _canonicalValue(value[key]),
    };
  }
  if (value case List<Object?>()) {
    return [for (final item in value) _canonicalValue(item)];
  }
  return value;
}

File: julia/test/semantic_index_runtime_projection_test.jl
# FUTURE-PARITY-BACKLOG.10.6.6.2 — immutable observed runtime projection.

const JULIA_RUNTIME_PROJECTION_RESPONSE_DIGEST =
    "36897041c6f71b95b577ce7b38f42d3649c6adffc6c37c069944a90f6eb65887"

function _julia_runtime_projection_source()
    return read(
        joinpath(
            REPO_ROOT,
            "capability_conformance",
            "semantic_introspection",
            "runtime.spec",
        ),
        String,
    )
end

function _julia_runtime_projection_base()
    return semantic_index(
        _julia_runtime_projection_source();
        logical_name = "runtime.spec",
        source_detail_ceiling = SemanticSourceTextDetail,
    )
end

function _julia_runtime_projection_capture()
    compiled = _julia_runtime_observation_compile(
        _julia_runtime_projection_source(),
    )
    return last(_julia_runtime_observation_capture() do sink
        runtime_parse(
            LinkedSpecRuntimeEngine(compiled),
            JULIA_RUNTIME_OBSERVATION_INPUT;
            semantic_observation_sink = sink,
        )
    end)
end

function _julia_runtime_projection_slot(;
    contract_id = RUNTIME_SEMANTIC_OBSERVATION_CONTRACT,
    rule_label = "Top",
    target_rule = "Top",
    regex_index = 0,
    position = 1,
    input_identity = nothing,
    status = nothing,
)
    return RuntimeSemanticObservationEvent(
        contract_id = contract_id,
        event_kind = RuntimeSemanticRegexSlotSelected,
        rule_label = rule_label,
        target_rule = target_rule,
        regex_index = regex_index,
        position = position,
        input_identity = input_identity,
        status = status,
    )
end

function _julia_runtime_projection_result(;
    contract_id = RUNTIME_SEMANTIC_OBSERVATION_CONTRACT,
    rule_label = "Top",
    target_rule = nothing,
    regex_index = nothing,
    position = 2,
    input_identity = JULIA_RUNTIME_OBSERVATION_INPUT_IDENTITY,
    status = "succeeded",
)
    return RuntimeSemanticObservationEvent(
        contract_id = contract_id,
        event_kind = RuntimeSemanticRuleResult,
        rule_label = rule_label,
        target_rule = target_rule,
        regex_index = regex_index,
        position = position,
        input_identity = input_identity,
        status = status,
    )
end

function _julia_runtime_projection_events()
    return RuntimeSemanticObservationEvent[
        _julia_runtime_projection_slot(),
        _julia_runtime_projection_slot(regex_index = 1, position = 2),
        _julia_runtime_projection_result(),
    ]
end

function _julia_runtime_projection_error(index, observation)
    return try
        with_execution_observation(index, observation)
        nothing
    catch error
        error
    end
end

@testset "Julia immutable semantic runtime observation projection" begin
    @testset "derived snapshot matches the twentieth typed and neutral digest" begin
        base = _julia_runtime_projection_base()
        events = _julia_runtime_projection_capture()
        query_case = _semantic_query_kernel_case("runtime_events")
        request = _semantic_query_kernel_request(query_case["request"])
        raw_request = deepcopy(query_case["request"])

        @test events == _julia_runtime_projection_events()
        @test !semantic_snapshot(base).has_execution
        @test isempty(semantic_query(base, request).records)
        derived = with_execution_observation(base, events)
        @test semantic_snapshot(derived).has_execution
        @test occursin("has_execution=true", sprint(show, derived))
        typed = semantic_query(derived, request)
        neutral = semantic_query_neutral(derived, raw_request)

        @test typed == neutral
        @test [record.id for record in typed.records] == [
            "execution:0",
            "event:execution:0:0",
            "event:execution:0:1",
            "event:execution:0:2",
        ]
        @test _semantic_query_kernel_digest(typed) ==
              JULIA_RUNTIME_PROJECTION_RESPONSE_DIGEST
        @test _semantic_query_kernel_digest(typed) ==
              query_case["expected"]["response_sha256"]

        relation_request = SemanticQuery(
            SemanticQueryRelationsOperation;
            subjects = ("execution:0",),
            relation_kinds = ("observed_as",),
            source = SemanticQuerySource(detail = SemanticSourceIdentityDetail),
        )
        relations = semantic_query(derived, relation_request)
        @test [relation.id for relation in relations.relations] == [
            "relation:observed_as:execution:0:event:execution:0:0:0",
            "relation:observed_as:execution:0:event:execution:0:1:1",
            "relation:observed_as:execution:0:event:execution:0:2:2",
        ]
        @test [collect(relation.evidence_ids) for relation in relations.relations] == [
            ["regex:rule:Top:0"],
            ["regex:rule:Top:1"],
            ["rule:Top"],
        ]

        events[1] = _julia_runtime_projection_slot(position = 999)
        detached = to_json(typed)
        detached["records"][1]["facts"]["status"] = "mutated"
        @test _semantic_query_kernel_digest(semantic_query(derived, request)) ==
              JULIA_RUNTIME_PROJECTION_RESPONSE_DIGEST
        @test !semantic_snapshot(base).has_execution
        @test isempty(semantic_query(base, request).records)
    end

    @testset "malformed foreign reordered and duplicate-final events reject" begin
        base = _julia_runtime_projection_base()
        valid = _julia_runtime_projection_events()
        cases = Any[
            ("not a vector", tuple(valid...)),
            ("non-event", Any[valid[1], "host", valid[3]]),
            ("empty", RuntimeSemanticObservationEvent[]),
            ("missing result", valid[1:2]),
            ("duplicate final", [valid..., valid[end]]),
            ("reordered final", [valid[1], valid[3], valid[2]]),
            (
                "foreign contract",
                [
                    _julia_runtime_projection_slot(contract_id = "future-contract"),
                    valid[2],
                    valid[3],
                ],
            ),
            (
                "missing rule label",
                [_julia_runtime_projection_slot(rule_label = ""), valid[2], valid[3]],
            ),
            (
                "missing slot target",
                [_julia_runtime_projection_slot(target_rule = nothing), valid[2], valid[3]],
            ),
            (
                "empty slot target",
                [_julia_runtime_projection_slot(target_rule = ""), valid[2], valid[3]],
            ),
            (
                "missing slot index",
                [_julia_runtime_projection_slot(regex_index = nothing), valid[2], valid[3]],
            ),
            (
                "negative slot index",
                [_julia_runtime_projection_slot(regex_index = -1), valid[2], valid[3]],
            ),
            (
                "negative slot position",
                [_julia_runtime_projection_slot(position = -1), valid[2], valid[3]],
            ),
            (
                "slot carries input identity",
                [
                    _julia_runtime_projection_slot(
                        input_identity = JULIA_RUNTIME_OBSERVATION_INPUT_IDENTITY,
                    ),
                    valid[2],
                    valid[3],
                ],
            ),
            (
                "slot carries result status",
                [_julia_runtime_projection_slot(status = "succeeded"), valid[2], valid[3]],
            ),
            (
                "result carries slot state",
                [
                    valid[1],
                    valid[2],
                    _julia_runtime_projection_result(
                        target_rule = "Top",
                        regex_index = 0,
                    ),
                ],
            ),
            (
                "foreign slot",
                [_julia_runtime_projection_slot(target_rule = "Missing"), valid[2], valid[3]],
            ),
            (
                "foreign result rule",
                [valid[1], valid[2], _julia_runtime_projection_result(rule_label = "Missing")],
            ),
            (
                "failed result",
                [valid[1], valid[2], _julia_runtime_projection_result(status = "failed")],
            ),
            (
                "missing result status",
                [valid[1], valid[2], _julia_runtime_projection_result(status = nothing)],
            ),
            (
                "negative result position",
                [valid[1], valid[2], _julia_runtime_projection_result(position = -1)],
            ),
            (
                "missing input identity",
                [valid[1], valid[2], _julia_runtime_projection_result(input_identity = nothing)],
            ),
            (
                "malformed input identity",
                [
                    valid[1],
                    valid[2],
                    _julia_runtime_projection_result(
                        input_identity = "input:sha256:xyz",
                    ),
                ],
            ),
            (
                "uppercase input identity",
                [
                    valid[1],
                    valid[2],
                    _julia_runtime_projection_result(
                        input_identity = string(
                            "input:sha256:",
                            uppercase(last(split(
                                JULIA_RUNTIME_OBSERVATION_INPUT_IDENTITY,
                                ':',
                            ))),
                        ),
                    ),
                ],
            ),
        ]
        @test length(cases) == 24
        for (label, observation) in cases
            captured = _julia_runtime_projection_error(base, observation)
            @test captured isa SemanticIndexError
            @test captured.stage == "execution_observation"
            @test captured.code == "semantic_index_invalid_observation"
            @test !isempty(captured.message)
            @test !isempty(label)
        end
    end

    @testset "selecting rule must own the selected slot relation" begin
        unrelated = semantic_index(
            """
Top::
 /a/ -> Top[0] { return("top") }

Other::
 /b/ -> Other[0] { return("other") }
""";
            logical_name = "unrelated.spec",
            source_detail_ceiling = SemanticSourceTextDetail,
        )
        events = RuntimeSemanticObservationEvent[
            _julia_runtime_projection_slot(rule_label = "Top", target_rule = "Other"),
            _julia_runtime_projection_result(rule_label = "Top"),
        ]
        captured = _julia_runtime_projection_error(unrelated, events)
        @test captured isa SemanticIndexError
        @test occursin(
            "does not select regex slot 'Other[0]'",
            captured.message,
        )
    end

    @testset "already observed and failed bases cannot derive again" begin
        valid = _julia_runtime_projection_events()
        derived = with_execution_observation(_julia_runtime_projection_base(), valid)
        repeated = _julia_runtime_projection_error(derived, valid)
        @test repeated isa SemanticIndexError
        @test repeated.stage == "execution_observation"
        @test repeated.code == "semantic_index_invalid_observation"

        failed = semantic_index(
            "Top:: Missing\n";
            logical_name = "failed.spec",
            source_detail_ceiling = SemanticSourceSpanDetail,
        )
        @test semantic_snapshot(failed).state == SemanticFailedCompilationSnapshotState
        failed_error = _julia_runtime_projection_error(failed, valid)
        @test failed_error isa SemanticIndexError
        @test failed_error.stage == "execution_observation"
        @test failed_error.code == "semantic_index_invalid_observation"
    end

    @testset "public derivation remains projection-only" begin
        @test :with_execution_observation in names(LinkedSpecJulia)
        implementation = read(
            joinpath(
                REPO_ROOT,
                "julia",
                "src",
                "semantic",
                "SemanticRuntimeProjection.jl",
            ),
            String,
        )
        for forbidden in (
            "parse_spec(",
            "validate_spec(",
            "compile_spec(",
            "runtime_parse(",
            "runtime_execute(",
            "LinkedSpecRuntimeEngine(",
            "LinkedSpecTraceEmitter(",
            "semantic_observation_sink",
            "SHA.sha256(",
            "ENV[",
            "read(",
            "open(",
        )
            @test !occursin(forbidden, implementation)
        end
    end
end
